BibBuddy/app/src/main/java/de/bibbuddy/BookFragment.java

```diff
@@ -203,80 +240,169 @@ public void onClick(DialogInterface dialog, int which) {
+                    + UriUtils.getUriFileName(getActivity(), uri), Toast.LENGTH_LONG).show();
+
+              } else {
+                showDialogNonBibFile();
+              }
+            }
+          }
+        }
+      });
+
+  private void handleImport(Uri uri) {
+    String bibText = readBibFile(uri);
+
+    List<String> nonRedundantBibItems
+        = importBibTex.getNonRedundantBibItems(bibText);
+
+    Book book;
+    for (int i = 0; i < nonRedundantBibItems.size(); i++) {
+
+      if (nonRedundantBibItems.get(i).startsWith(BibTexKeys.BOOK_TAG)) {
+        importBibTex.parseBibItem(nonRedundantBibItems.get(i));
+        book = importBibTex.importBook();
+        addImportedBook(book, importBibTex.parseAuthorNames());
+        importBibTex.importBibNote(noteDao, book);
+      }
+
+    }
+
+  }
+
+  private void addImportedBook(Book book, List<Author> authorList) {
+
+    if (bibTagValue.get(BibTexKeys.ANNOTE).isEmpty()) {
+      bookModel.addBook(book, authorList);
+
     } else {
-      // if the user has already allowed access to device external storage
-      exportBibTex.createBibFile();
-      exportBibTex.writeBibFile(exportBibTex.getBibDataFromShelf(shelfId, bookDao, noteDao));
-
-      Toast.makeText(getContext(),
-          getString(R.string.exported_file_stored_in) + '\n'
-              + File.separator + StorageKeys.DOWNLOAD_FOLDER + File.separator
-              + shelfName + StorageKeys.BIB_FILE_TYPE, Toast.LENGTH_LONG).show();
+      bookModel.addImportedBook(book, authorList);
     }
+
+    adapter.setBookList(bookModel.getCurrentBookList());
+    adapter.notifyDataSetChanged();
+    updateEmptyView(bookModel.getCurrentBookList());
   }
 
-  private void requestStoragePermission() {
-    if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
-        Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+
+  private String readBibFile(Uri uri) {
+    try {
+      return importBibTex.readTextFromUri(uri);
+
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    return null;
+  }
+
+  private void filePicker()  {
+    filePickerIntent = new Intent(Intent.ACTION_GET_CONTENT);
+    filePickerIntent.setType("*/*");
+    filePickerActivityResultLauncher.launch(filePickerIntent);
+  }
+
+  private void showDialogNonBibFile() {
+    AlertDialog.Builder nonBibFileAlertDialog = new AlertDialog.Builder(context);
+
+    nonBibFileAlertDialog.setTitle(R.string.import_non_bib_file);
+    nonBibFileAlertDialog.setMessage(R.string.import_non_bib_file_description);
+
+    nonBibFileAlertDialog.setPositiveButton(R.string.ok,
+        (dialog, which) -> filePicker());
+
+    nonBibFileAlertDialog.setNegativeButton(R.string.cancel,
+        (dialog, which) -> dialog.dismiss());
+
+    nonBibFileAlertDialog.create().show();
+  }
+
+  private void checkStoragePermission() {
+    // if the permissions are granted
+    if (ContextCompat.checkSelfPermission(getContext(), storageManifestPermissions[0])
```

Comment:
Instead of using `storageManifestPermissions[0]` you should directly use the permission name because without looking at the `storageManifestPersmissions` Array I don't know what it contains at index 0.

Reply:
I will add a comment in those lines because so far we used an array of filter criteria in SearchModel class, that also uses indexes and not direct value names.